src/mon/OSDMonitor.cc

```diff
@@ -4551,6 +4551,7 @@ int OSDMonitor::crush_rename_bucket(const string& srcname,
     return ret;
 
   pending_inc.crush.clear();
+  newcrush.cleanup_classes();
   newcrush.encode(pending_inc.crush, mon->get_quorum_con_features());
```

Comment:
Hum. The
```
      cleanup_classes();
      populate_classes();
```
in CrushWrapper::decode should actually be
```
      cleanup_classes();
```
and we can get rid of mots the other calls in OSDMonitor. 

case a) A non device classes aware client gets a crushmap from a device class aware mon, modify it, encode it, send it back to a device class aware mon. OK

case b) A non device class aware client gets a crushmap from a device aware mon including generated buckets, and calls a non device class aware crushtool to decompile it. It will fail to recompile because of the special bucket names. A better error message suggesting an upgrade should be displayed in this case.

case C) A device class aware client sends a crushmap to a non device class aware mon (maybe because it is the leader of a cluster where an upgrade is in progress) the mon will be able to use the crushmap exactly as a device class aware mon would. Only it will never trim unused buckets with special names.

The ```crush.populate_classes();``` in CrushCompiler::decompile should be removed.
The ```  crush.populate_classes();``` in CrushCompiler::parse_crush should really be cleanup_classes since populate_classes() is called before parsing a rule.
All cleanup_classes from OSDMonitor should be removed since it is called by decode anyway.
The cleanup_classes from crushtool.cc should be removed since it is is called from CrushCompiler::parse_crush and from CrushCompiler::decompile

I'll do these changes if they makes sense to you. Or if it's all too confusing and implementing them will help clarify ;-)